File: week6/sentimental-credit/credit.py
from cs50 import get_string
import re

numero = get_string("cc number: ")


def label(n):

    if re.search("^(34|37)[0-9]{13}$", numero):
        return "AMEX"
    elif re.search("^(51|52|53|54|55)[0-9]{14}$", numero):
        return "MASTERCARD"
    elif re.search("^(4)[0-9]{12,15}$", numero):
        return "VISA"
    else:
        return "INVALID"


def verificador(n):

    l = len(n)

    odd = 0
    even = 0

    for idx, val in enumerate(numero):

        posicao = l - idx

        x = int(val)

        if posicao % 2 == 0:
            even += int((x * 2) / 10) + ((x * 2) % 10)
        else:
            odd += x

    if (odd + even) % 10 == 0:
        return True
    else:
        return False


if verificador(numero):
    print(label(numero))
else:
    print("INVALID")

File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

string label(long numero);
bool verificador(long numero);

int main(void)
{

    long ccnumber = get_long("Numero: ");

    if (verificador(ccnumber))
    {
        printf("%s", label(ccnumber));
    }
    else
    {
        printf("INVALID\n");
    }
}

string label(long numero)
{
    char str[20];
    sprintf(str, "%ld", numero);

    int tamanho = strlen(str);

    char fdigit = str[0];
    char sdigit = str[1];

    if (fdigit == '4' && (tamanho == 13 || tamanho == 16))
    {
        return "VISA\n";
    }
    else if (((fdigit == '3' && sdigit == '4') || (fdigit == '3' && sdigit == '7')) && (tamanho == 15))
    {
        return "AMEX\n";
    }
    else if (((fdigit == '5' && sdigit == '1') || (fdigit == '5' && sdigit == '2') || (fdigit == '5' && sdigit == '3') ||
              (fdigit == '5' && sdigit == '4') || (fdigit == '5' && sdigit == '5')) &&
             (tamanho == 16))
    {
        return "MASTERCARD\n";
    }
    else
    {
        return "INVALID\n";
    }
}

bool verificador(long numero)
{
    char str[20];
    sprintf(str, "%ld", numero);
    int tamanho = strlen(str);

    int odd = 0;
    int even = 0;

    for (int i = 0; i < tamanho; i++)
    {
        int posicao = tamanho - i;

        int x = str[i] - '0';

        if ((posicao % 2) == 0)
        {
            int additional = x * 2;
            even += (additional / 10) + (additional % 10);
        }
        else
        {
            odd += x;
        }
    }

    if ((even + odd) % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}
